HTTP 400 and NPE in HttpParser for blank header value in Jetty 12.x

**Jetty Version**
12.0.5

**Jetty Environment**
`ee10` but not relevant in this case

**Java Version**
21.0.2 (Corretto)

**Question**
After upgrading from 11.x to 12.x noticed a difference in header parsing. One of our tests for bad auth header values is now failing inside Jetty header parsing without reaching our handler. Below is a minimal reproducer with Jetty 12.0.5 and Apache HTTP client:

```java
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.IOException;
import org.apache.http.HttpVersion;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.HttpClients;
import org.eclipse.jetty.server.Handler;
import org.eclipse.jetty.server.Request;
import org.eclipse.jetty.server.Response;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.ErrorHandler;
import org.eclipse.jetty.util.Callback;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.junit.platform.commons.logging.Logger;
import org.junit.platform.commons.logging.LoggerFactory;

public class Jetty12HeaderTest {
  static final Logger log = LoggerFactory.getLogger(Jetty12HeaderTest.class);

  @ValueSource(strings = {"", " "})
  @ParameterizedTest
  void it_does_not_fail_on_blank_header_value(String headerValue) throws Exception {
    var server = new Server(0);
    server.setHandler(new Handler.Abstract() {
      @Override
      public boolean handle(final Request request, final Response response, final Callback callback) throws Exception {
        response.setStatus(200);
        callback.succeeded();
        return true;
      }
    });
    server.setErrorHandler(new ErrorHandler() {
      @Override
      protected void generateResponse(final Request request, final Response response, final int code,
          final String message, final Throwable cause, final Callback callback) throws IOException {
        log.error(cause, () -> "Error handler called with code " + code + " and message " + message);
        callback.succeeded();
      }
    });
    server.start();
    try (var client = HttpClients.createDefault()) {
      var request = new HttpPost(server.getURI().toString());
      request.setHeader("Authorization", headerValue);
      request.setProtocolVersion(HttpVersion.HTTP_1_0);
      try (var response = client.execute(request)) {
        assertEquals(200, response.getStatusLine().getStatusCode(),
            () -> "Response status was " + response.getStatusLine());
      }
    } finally {
      server.stop();
    }
  }
}
```

I expect Jetty to pass the blank `" "` header value just like the empty `""` header value to the hander and let the handler decide what to do. This was the behavior in Jetty 11, but in Jetty 12 the blank value will throw a NPE inside Jetty header parser and result in HTTP 400 response. 

```
Jan 18, 2024 1:48:34 P.M. Jetty12HeaderTest$2 generateResponse
SEVERE: Error handler called with code 400 and message Bad Request
org.eclipse.jetty.http.BadMessageException: 400: Bad Request
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1652)
	at org.eclipse.jetty.server.internal.HttpConnection.parseRequestBuffer(HttpConnection.java:580)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:404)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "String.length()" because "valueString" is null
	at org.eclipse.jetty.http.HttpParser$FieldCache.cacheable(HttpParser.java:2131)
	at org.eclipse.jetty.http.HttpParser.parsedHeader(HttpParser.java:1091)
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1268)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1543)
	... 9 more
```

Could you please check if this is a bug or an expected behavior change in header parsing?

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
@@ -2128,7 +2128,7 @@ public class HttpParser
 
         public boolean cacheable(HttpHeader header, String valueString)
         {
-            return isEnabled() && header != null && valueString.length() <= _size;
+            return isEnabled() && header != null && valueString != null && valueString.length() <= _size;
         }
 
         private void prepare()

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
@@ -557,12 +557,23 @@ public class HttpParserTest
         ByteBuffer buffer = BufferUtil.toBuffer(
             "GET / HTTP/1.0" + eoln +
                 "Host: localhost" + eoln +
-                "Name0: " + eoln +
+                "Name0:  " + eoln +
                 "Name1:" + eoln +
+                "Authorization:  " + eoln +
+                "Authorization:" + eoln +
                 eoln);
 
-        HttpParser.RequestHandler handler = new Handler();
+        HttpParser.RequestHandler handler = new Handler()
+        {
+            @Override
+            public void badMessage(HttpException failure)
+            {
+                ((Throwable)failure).printStackTrace();
+                super.badMessage(failure);
+            }
+        };
         HttpParser parser = new HttpParser(handler);
+        parser.setHeaderCacheSize(1024);
         parseAll(parser, buffer);
 
         assertTrue(_headerCompleted);
@@ -576,7 +587,11 @@ public class HttpParserTest
         assertEquals("", _val[1]);
         assertEquals("Name1", _hdr[2]);
         assertEquals("", _val[2]);
-        assertEquals(2, _headers);
+        assertEquals("Authorization", _hdr[3]);
+        assertEquals("", _val[3]);
+        assertEquals("Authorization", _hdr[4]);
+        assertEquals("", _val[4]);
+        assertEquals(4, _headers);
     }
 
     @ParameterizedTest

